mono/mini/mini-llvm.c

```diff
@@ -9623,6 +9675,83 @@ emit_aot_file_info (MonoLLVMModule *module)
 	}
 }
 
+typedef struct {
+	LLVMValueRef lmethod;
+	int argument;
+} NonnullPropWorkItem;
+
+static void
+mono_llvm_propagate_nonnull_func (GHashTable *all_directly_called, MonoLLVMModule *module, LLVMValueRef root_lmethod, int argument)
+{
+	// When we first traverse the mini IL, we mark the things that are
+	// nonnull (the roots). Then, for all of the methods that can be specialized, we
+	// see if their call sites have nonnull attributes. 
+
+	// If so, we mark the function's param. This param has uses to propagate
+	// the attribute to. This propagation can trigger a need to mark more attributes
+	// non-null, and so on and so forth.
+	GSList *queue = NULL;
+
+	NonnullPropWorkItem *root = g_malloc (sizeof (NonnullPropWorkItem));
+	root->lmethod = root_lmethod;
+	root->argument = argument;
+
+	queue = g_slist_prepend (queue, root);
+
+	while (queue) {
+		NonnullPropWorkItem *current = (NonnullPropWorkItem *) queue->data;
+		queue = queue->next;
+		g_assert (current->argument < LLVMCountParams (current->lmethod));
+
+		mono_llvm_set_func_nonnull_arg (current->lmethod, current->argument);
+		LLVMValueRef caller_argument = LLVMGetParam (current->lmethod, current->argument);
+
+		GSList *calls = mono_llvm_calls_using (caller_argument);
+		for (GSList *cursor = calls; cursor != NULL; cursor = cursor->next) {
+			LLVMValueRef lcall = (LLVMValueRef) cursor->data;
+			LLVMValueRef callee_lmethod = LLVMGetCalledValue (lcall);
+
+			// If this wasn't a direct call, this lookup won't find a MonoMethod. By limiting ourselves to
+			// direct calls, we entire that the arity of the call site and the arity of the function are identical
+			MonoMethod *callee_method = (MonoMethod *) g_hash_table_lookup (all_directly_called, callee_lmethod);
+			if (!mono_aot_can_specialize (callee_method))
+				continue;
+
+			// Decrement number of nullable refs at that func's arg offset
+			GArray *call_site_union = (GArray *) g_hash_table_lookup (module->method_to_call_info, callee_method);
+			int max_params = LLVMCountParams (callee_lmethod);
+			if (call_site_union->len != max_params) {
+				mono_llvm_dump_value (callee_lmethod);
+				g_assert_not_reached ();
+			}
+
+			LLVMValueRef *operands = mono_llvm_call_args (lcall);
+			for (int call_argument = 0; call_argument < max_params; call_argument++) {
+				if (caller_argument == operands [call_argument]) {
+
+					gint32 *nullable_count = &g_array_index (call_site_union, gint32, call_argument);
+					g_assert (*nullable_count > 0);
+					*nullable_count = *nullable_count - 1;
+
+					if (*nullable_count == 0) {
+						NonnullPropWorkItem *item = g_malloc (sizeof (NonnullPropWorkItem));
+						item->lmethod = callee_lmethod;
+						item->argument = call_argument;
+						queue = g_slist_prepend (queue, item);
+					}
+				}
+			}
+			g_free (operands);
+
+			g_hash_table_insert (module->method_to_call_info, callee_method, call_site_union);
+		}
+		g_free (calls);
```

Comment:
This is only freeing the head of the list.  should be `g_slist_free`.

Reply:
Oh that's right. This was just a C array before. Will fix.